<a href="https://colab.research.google.com/github/KosmanAya/PP2/blob/master/python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas. Домашнее задание

В домашнем задании есть обязательные и дополнительные задачи.

*Обязательные задачи* нужно сделать, чтобы проверить, что вы действительно усвоили материал модуля. Они обязательны к загрузке на проверку преподавателю.

*Дополнительные задачи* для тех, кто чувствует, что нужно ещё немного потренироваться и порешать задачки по работе с Pandas. Вы также можете загрузить их на проверку преподавателю и получить по ним обратную связь.

Удачи!



---



---



**Цели домашнего задания:**
1. Научиться работать с описательными статистиками набора данных.
2. Научиться выбирать данные по условиям.
3. Научиться объединять несколько датафреймов в один.


**Что входит в домашнее задание:**
1. *Обязательные задачи*:

17 задач с использованием:
- набора рекламных данных, показывающих, щелкнул ли конкретный пользователь интернета рекламу на веб-сайте компании или нет,
- дополнительного набора данных, описывающего пользователей, посещающих веб-сайт компании.
2. *Дополнительные задачи*:

3 задачи для отработки исследования данных.


**Критерии оценки:**
- программа выдаёт верный ответ на заданном наборе данных;
- код читабелен: переменным даны осмысленные названия, отступы и правила расстановки пробелов соблюдены.

**Как отправить задание на проверку**

Выполните задачи, сохраните изменения при помощи опции Save and Checkpoint из вкладки меню File или кнопки Save and Checkpoint на панели инструментов. Отправьте через форму на платформе итоговый файл Jupyter Notebook (в формате .ipynb) или ссылку на него.



---


---



**Обязательные задачи**

В течение всего домашнего задания вы будете работать с несколькими наборами данных:
1. Набор рекламных данных, показывающих, щёлкнул ли конкретный пользователь рекламу на сайте компании или нет. Состоит из трёх разных файлов: `advertising_1.csv`, `advertising_2.csv`, `advertising_3.csv`.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Daily Time Spent on Site**: Среднее ежедневное время нахождения пользователя на сайте в минутах.
- **Daily Internet Usage**: Среднее ежедневное время нахождения пользователя в интернете в минутах.
- **Ad Topic Line**: Заголовок рекламного объявления.
- **Clicked on Ad**: Кликнул пользователь на рекламу или нет.

Дополнительный к рекламным данным датасет, который описывает пользователей, посещающих сайт компании. Состоит из одного файла: users.csv.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Age**: Возраст пользователя.
- **Male**: Является ли пользователь мужчиной.
- **City**: Город пользователя.
- **Country**: Страна пользователя.
- **Area Income**: Средний доход в месте жительства пользователя.


***Задача 1***

Загрузите таблицу из файла `advertising_1.csv` и сохраните её в датафрейм `adv1_df`. Укажите в качестве индекса столбец `Number`. Выведите на экран первые десять строк.


In [18]:
# Ваш код здесь
import pandas as pd
with open('advertising_1.csv') as file:
  adv1_df=pd.read_csv('advertising_1.csv', index_col='Number', header=0)
adv1_df.head(10)


Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
11                         88.91                208.36   
12                           NaN                131.76   
13                         74.53                   NaN   
14                         69.88                183.82   
15                         47.64                122.02   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
5          Cloned 5thgeneration orchestration              0  
6          Monitored national standardization              0  
8       Triple-buffered reciprocal time-frame              0  
9               Robust logistical utilization              0  
10            Sharable client-driven software              0  
11                 Enhanced dedicated support              0  
12                   Reactive local challenge              1  
13             Configurable coherent function              0  
14         Mandatory homogeneous architecture              0  
15             Centralized neutral neural-net              1

***Задача 2***

Выведите размер датафрейма `adv1_df`.

In [22]:
# Ваш код здесь
adv1_df.shape

(500, 4)

***Задача 3***

Выведите cреднее ежедневное время нахождения в интернете пользователя под номером 8.

In [25]:
# Ваш код здесь
adv1_df.loc[8, 'Daily Time Spent on Site']


74.15

***Задача 4***

Загрузите таблицу из файла `advertising_2.csv` и сохраните её в датафрейм `adv2_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран строки для пользователей с номерами с 533-го по 536-й.

In [5]:
# Ваш код здесь
import pandas as pd
with open('advertising_2.csv') as file:
  adv2_df=pd.read_csv('advertising_1.csv', index_col='Number', header=0)
adv2_df.loc[519:520]

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
519                          NaN                250.03   
520                        42.94                130.40   

                             Ad Topic Line  Clicked on Ad  
Number                                                     
519     Future-proofed modular utilization              0  
520      Synergistic dynamic orchestration              1

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

***Задача 5***

Выведите сводную таблицу описательных статистик для датафрейма `adv2_df`, вызвав необходимый метод на всём датафрейме.

Глядя на вывод, напишите, столбец с каким названием не содержит незаполненных значений.

*Подсказка:*  В сводной таблице не содержится информации о количестве строк всего датафрейма, поэтому для сначала распечатайте общее количество строк датафрейма.

*Рекомендация:* Кроме изученного метода `describe` для описания датафрейма и получения информации о пропущенных значениях полезен метод [`info`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html).

In [ ]:
# Ваш код здесь
adv2_df.describe()
adv2_df

In [ ]:
# Ваш код здесь
print('Не содержит незаполненных значений столбец:', ваше_значение)

***Задача 6***

Добавьте к датафрейму `adv1_df` в конец датафрейм `adv2_df` с помощью метода `append`.

Сохраните объединенный датафрейм в новой переменной `adv12_df` и распечатайте его.

In [19]:
adv12_df = adv1_df.append(adv2_df)
adv12_df

<ipython-input-19-74196f96684b>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  adv12_df = adv1_df.append(adv2_df)


Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
...                          ...                   ...   
525                        46.61                136.18   
526                        63.26                120.46   
527                        79.16                202.90   
528                        67.94                128.16   
529                        79.91                230.18   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
5          Cloned 5thgeneration orchestration              0  
6          Monitored national standardization              0  
8       Triple-buffered reciprocal time-frame              0  
9               Robust logistical utilization              0  
10            Sharable client-driven software              0  
...                                       ...            ...  
525       Progressive intermediate throughput              1  
526             Customizable holistic archive              1  
527           Compatible intermediate concept              0  
528      Assimilated next generation firmware              1  
529        Total zero administration software              0  

[1000 rows x 4 columns]

***Задача 7***

Для объединённого датафрейма `adv12_df` распечатайте среднее время, проводимое пользователем на сайте компании.

In [20]:
# Ваш код здесь
adv12_df['Daily Internet Usage'].mean()


179.49316964285714

***Задача 8***

Из объединённого датафрейма `adv12_df` выберите и распечатайте строки только с теми пользователями, для которых неизвестно среднее ежедневное время, проводимое пользователем на сайте компании.

In [ ]:
# Ваш код здесь

***Задача 9***

Загрузите таблицу из файла `advertising_3.csv` и сохраните её в датафрейм `adv3_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран столбцы `Ad Topic Line` и `Clicked on Ad`.

In [14]:
# Ваш код здесь
import pandas as pd
with open ('advertising_3.csv') as file:
  adv3_df = pd.read_csv('advertising_3.csv', index_col='Number', header=0)
adv3_df.loc[:,['Ad Topic Line','Clicked on Ad']]

Ad Topic Line  Clicked on Ad
Number                                                             
740                 Distributed bifurcated challenge              0
741       Customizable zero-defect Internet solution              0
742     Self-enabling zero administration neural-net              1
744           Open-architected system-worthy ability              0
745               Quality-focused maximized extranet              1
...                                              ...            ...
999                    Front-line bifurcated ability              1
1000                   Fundamental modular algorithm              1
1001                 Grass-roots cohesive monitoring              1
1002                    Expanded intangible solution              1
1003            Proactive bandwidth-monitored policy              0

[251 rows x 2 columns]

***Задача 10***

Добавьте к датафрейму `adv12_df` в конец датафрейм `adv3_df` с помощью метода `concat`. Явно пропишите значение параметра `axis`.

Сохраните объединённый датафрейм в новой переменной `adv123_df` и распечатайте его.

In [26]:
# Ваш код здесь
import pandas as pd
adv123_df=pd.concat([adv12_df, adv3_df], axis=0)
adv123_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
...                          ...                   ...   
999                        43.70                173.01   
1000                       72.97                208.58   
1001                       51.30                   NaN   
1002                       51.63                120.37   
1003                       55.55                187.95   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
5          Cloned 5thgeneration orchestration              0  
6          Monitored national standardization              0  
8       Triple-buffered reciprocal time-frame              0  
9               Robust logistical utilization              0  
10            Sharable client-driven software              0  
...                                       ...            ...  
999             Front-line bifurcated ability              1  
1000            Fundamental modular algorithm              1  
1001          Grass-roots cohesive monitoring              1  
1002             Expanded intangible solution              1  
1003     Proactive bandwidth-monitored policy              0  

[1251 rows x 4 columns]

***Задача 11***

Для объединённого датафрейма `adv123_df` выберите только те строки, в которых пользователи кликнули на рекламное сообщение. Сохраните полученный результат в датафрейм `success_adv_df`, распечатайте размер нового датафрейма.

In [62]:
# Ваш код здесь
success_adv_df = adv123_df[adv123_df['Clicked on Ad']==1]
success_adv_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
12                           NaN                131.76   
15                         47.64                122.02   
17                         69.57                113.12   
19                           NaN                143.56   
20                         63.45                140.64   
...                          ...                   ...   
997                        69.17                123.62   
999                        43.70                173.01   
1000                       72.97                208.58   
1001                       51.30                   NaN   
1002                       51.63                120.37   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
12                   Reactive local challenge              1  
15             Centralized neutral neural-net              1  
17      Centralized content-based focus group              1  
19              Grass-roots coherent extranet              1  
20         Persistent demand-driven interface              1  
...                                       ...            ...  
997             Extended leadingedge solution              1  
999             Front-line bifurcated ability              1  
1000            Fundamental modular algorithm              1  
1001          Grass-roots cohesive monitoring              1  
1002             Expanded intangible solution              1  

[643 rows x 4 columns]

***Задача 12***

Загрузите таблицу с данными пользователей из файла `users.csv` и сохраните её в датафрейм `users_df`. Индекс не указывайте.

Распечатайте сам датафрейм и сводную таблицу со статистиками для него.

In [32]:
# Ваш код здесь
import pandas as pd
with open ('users.csv') as file:
  users_df = pd.read_csv('users.csv')
users_df
users_df.describe()





Number         Age        Male   Area Income
count   980.000000  860.000000  868.000000    903.000000
mean    504.485714   36.073256    0.500000  55010.631938
std     287.956480    8.795992    0.500288  13423.710892
min       5.000000   19.000000    0.000000  13996.500000
25%     255.750000   29.000000    0.000000  46947.570000
50%     503.500000   35.000000    0.500000  57032.360000
75%     753.250000   42.000000    1.000000  65537.990000
max    1004.000000   61.000000    1.000000  79484.800000

***Задача 13***

Глядя на статистики `users_df`, укажите значения, удовлетворяющие следующим условиям:
1. Возраст самого молодого пользователя сайта: *ваш_ответ*
2. Максимальный средний доход в месте жительства пользователя сайта: *ваш_ответ*
3. Средний возраст пользователя сайта: *ваш_ответ*
4. Количество заполненных значений для колонки, содержащей информацию о том, является пользователь мужчиной или нет: *ваш_ответ*
5. Минимальный номер пользователя: *ваш_ответ*

В этом задании не требуется выбирать по условиям и распечатывать значения, достаточно внимательно посмотреть на статистики и скопировать нужные значения согласно вопросам.

***Задача 14***

К полученному датафрейму с успешными рекламными сообщениями `success_adv_df` добавьте справа информацию из датафрейма с пользователями `users_df` и сохраните в датафрейм `success_full_df`. В полученном датафрейме должны остаться только те пользователи, номера которых есть в обоих датафреймах. Для этого воспользуйтесь методом `merge` и пропишите явно верные параметры.

Затем распечатайте датафрейм.

*Важно:*

Объединять датафреймы с помощью `merge` можно и по индексам, и по столбцам с разными названиями.
Так как общие значения содержатся в столбце `Number`, но в датафрейме `users_df` данный столбец является обычным столбцом, а в `success_adv_df` является индексом, перед тем как объединять, переведите столбец `Number` в `success_adv_df` в обычный столбец.

Для этого перед объединением пропишите:
```
success_adv_df = success_adv_df.reset_index()
```

[Подробнее о reset_index](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html)

In [65]:
# Ваш код здесь
import pandas as pd
success_adv_df = success_adv_df.reset_index()
success_full_df = pd.merge(success_adv_df, users_df, on='Number')
success_full_df = success_full_df.set_index('Number')
success_full_df

level_0  index  Daily Time Spent on Site  Daily Internet Usage  \
Number                                                                   
12            0      0                       NaN                131.76   
12          252    252                       NaN                131.76   
15            1      1                     47.64                122.02   
15          253    253                     47.64                122.02   
17            2      2                     69.57                113.12   
...         ...    ...                       ...                   ...   
997         638    638                     69.17                123.62   
999         639    639                     43.70                173.01   
1000        640    640                     72.97                208.58   
1001        641    641                     51.30                   NaN   
1002        642    642                     51.63                120.37   

                                Ad Topic Line  Clicked on Ad   Age  Male  \
Number                                                                     
12                   Reactive local challenge              1  48.0   1.0   
12                   Reactive local challenge              1  48.0   1.0   
15             Centralized neutral neural-net              1  49.0   NaN   
15             Centralized neutral neural-net              1  49.0   NaN   
17      Centralized content-based focus group              1  48.0   1.0   
...                                       ...            ...   ...   ...   
997             Extended leadingedge solution              1  40.0   NaN   
999             Front-line bifurcated ability              1  28.0   0.0   
1000            Fundamental modular algorithm              1  30.0   NaN   
1001          Grass-roots cohesive monitoring              1  45.0   1.0   
1002             Expanded intangible solution              1  51.0   NaN   

                    City                 Country  Area Income  
Number                                                         
12      Port Jefferybury               Australia     24593.33  
12      Port Jefferybury               Australia     24593.33  
15       West Brandonton                   Qatar     45632.51  
15       West Brandonton                   Qatar     45632.51  
17        West Katiefurt                   Egypt     51636.92  
...                  ...                     ...          ...  
997          New Patrick              Montenegro     66193.81  
999         Nicholasland                 Mayotte     63126.96  
1000           Duffystad                 Lebanon     71384.57  
1001         New Darlene  Bosnia and Herzegovina     67782.17  
1002       South Jessica                Mongolia     42415.72  

[630 rows x 11 columns]

***Задача 15***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, отобразите сводную таблицу со статистиками для строковых колонок (`Ad Topic Line`, `City`, `Country`).


In [67]:
# Ваш код здесь
success_full_df[['Ad Topic Line', 'City', 'Country']].describe()


Ad Topic Line          City    Country
count                        630           567        573
unique                       382           338        189
top     Reactive local challenge  Michelleside  Australia
freq                           2             4         10

***Задача 16***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите страну и количество вхождений этой страны в датафрейм.


In [73]:
# Ваш код здесь
if success_full_df_1=success_full_df[success_full_df['Clicked on Ad']==1]:

success_full_df_1

level_0  index  Daily Time Spent on Site  Daily Internet Usage  \
Number                                                                   
12            0      0                       NaN                131.76   
12          252    252                       NaN                131.76   
15            1      1                     47.64                122.02   
15          253    253                     47.64                122.02   
17            2      2                     69.57                113.12   
...         ...    ...                       ...                   ...   
997         638    638                     69.17                123.62   
999         639    639                     43.70                173.01   
1000        640    640                     72.97                208.58   
1001        641    641                     51.30                   NaN   
1002        642    642                     51.63                120.37   

                                Ad Topic Line  Clicked on Ad   Age  Male  \
Number                                                                     
12                   Reactive local challenge              1  48.0   1.0   
12                   Reactive local challenge              1  48.0   1.0   
15             Centralized neutral neural-net              1  49.0   NaN   
15             Centralized neutral neural-net              1  49.0   NaN   
17      Centralized content-based focus group              1  48.0   1.0   
...                                       ...            ...   ...   ...   
997             Extended leadingedge solution              1  40.0   NaN   
999             Front-line bifurcated ability              1  28.0   0.0   
1000            Fundamental modular algorithm              1  30.0   NaN   
1001          Grass-roots cohesive monitoring              1  45.0   1.0   
1002             Expanded intangible solution              1  51.0   NaN   

                    City                 Country  Area Income  
Number                                                         
12      Port Jefferybury               Australia     24593.33  
12      Port Jefferybury               Australia     24593.33  
15       West Brandonton                   Qatar     45632.51  
15       West Brandonton                   Qatar     45632.51  
17        West Katiefurt                   Egypt     51636.92  
...                  ...                     ...          ...  
997          New Patrick              Montenegro     66193.81  
999         Nicholasland                 Mayotte     63126.96  
1000           Duffystad                 Lebanon     71384.57  
1001         New Darlene  Bosnia and Herzegovina     67782.17  
1002       South Jessica                Mongolia     42415.72  

[630 rows x 11 columns]

***Задача 17***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных.


In [ ]:
# Ваш код здесь

***Задача 18***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных, они проводят в интернете в среднем больше 120 минут и их возраст меньше 30 лет.
Cохраните получившиеся строки в новый датафрейм `filtered_df`.


In [ ]:
# Ваш код здесь

***Задача 19***

Для получившегося датафрейма `filtered_df` выведите строки с данными пользователей, для которых либо не указано среднее время, проведённое на сайте, либо указанное время больше 55 минут. Cохраните получившиеся строки в новый датафрейм `filtered_df2`.


In [ ]:
# Ваш код здесь

**Дополнительные задачи**

***Задача 20***

Для датафрейма `adv123_df` выведите строки с данными, заголовок рекламного сообщения в которых включает слово `model`.

Для этого вам понадобится метод [`contains`](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html).

Cохраните получившиеся строки в новый датафрейм `model_df`.


In [78]:
# Ваш код здесь
model_df = adv123_df.loc[adv123_df['Ad Topic Line'].str.contains('model')]
model_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
62                         44.98                129.31   
119                        69.00                   NaN   
202                        78.57                   NaN   
211                        76.90                212.67   
314                          NaN                161.16   
350                        65.77                   NaN   
501                        81.10                216.49   
62                         44.98                129.31   
119                        69.00                   NaN   
202                        78.57                   NaN   
211                        76.90                212.67   
314                          NaN                161.16   
350                        65.77                   NaN   
501                        81.10                216.49   
823                        81.98                212.88   
995                        35.79                165.62   

                              Ad Topic Line  Clicked on Ad  
Number                                                      
62                   Organized global model              1  
119                 Virtual composite model              0  
202           Exclusive client-driven model              0  
211           Synchronized multimedia model              0  
314              Extended interactive model              1  
350     Function-based fault-tolerant model              0  
501                Front-line dynamic model              0  
62                   Organized global model              1  
119                 Virtual composite model              0  
202           Exclusive client-driven model              0  
211           Synchronized multimedia model              0  
314              Extended interactive model              1  
350     Function-based fault-tolerant model              0  
501                Front-line dynamic model              0  
823                Up-sized maximized model              0  
995          Enterprise-wide tangible model              1

***Задача 21***

Для получившегося датафрейма `model_df` выведите строки с данными пользователей, которые проводят на сайте больше среднего времени по всем пользователям в данном датафрейме. Cохраните получившиеся строки в новый датафрейм `model_popular_df`.


In [79]:
# Ваш код здесь
if 'Daily Time Spent on Site'>15.26:
model_df.describe()

Daily Time Spent on Site  Daily Internet Usage  Clicked on Ad
count                 14.000000             10.000000      16.000000
mean                  67.886429            181.776000       0.312500
std                   15.265900             36.366771       0.478714
min                   35.790000            129.310000       0.000000
25%                   65.770000            161.160000       0.000000
50%                   72.950000            189.145000       0.000000
75%                   78.570000            212.827500       1.000000
max                   81.980000            216.490000       1.000000

***Задача 22***

Для объединённого датафрейма `model_popular_df` выведите строки с данными пользователей, номера которых равны 501, 712 и 823.

Для этого вам понадобится метод [`isin`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isin.html).

*Подсказка:* Не забудьте, что номера пользователей находятся не в обычной колонке, а в колонке, являющейся индексом, поэтому вам необходимо получить значения индекса, а не обычной колонки.

In [ ]:
# Ваш код здесь